Documentation/exposing-prometheus-and-alertmanager.md

```diff
@@ -0,0 +1,291 @@
+
+## NodePort
+
+The easiest way to expose Prometheus or Alertmanager is to use a `Service` of
+type `NodePort`.
+
+A simple Prometheus object's manifest could look like this:
+
+```
+apiVersion: "monitoring.coreos.com/v1alpha1"
+kind: "Prometheus"
+metadata:
+  name: "main"
+spec:
+  replicas: 1
+  version: v1.5.0
+  resources:
+    requests:
+      memory: 400Mi
+```
+
+All Prometheus `Pod`s are labelled with `prometheus: <prometheus-name>`, as the
+Prometheus object's name is `main`, the selector ends up being `prometheus:
+main`. Meaning, the respective manifest for the `Service` would look like this:
+
+```
+apiVersion: v1
+kind: Service
+metadata:
+  name: prometheus-main
+spec:
+  type: NodePort
+  ports:
+  - name: web
+    nodePort: 30900
+    port: 9090
+    protocol: TCP
+    targetPort: web
+  selector:
+    prometheus: main
+```
+
+After creating a `Service` with the above manifest, the web UI of Prometheus
+will be accessible by browsing to any of the worker nodes using
+`http://node-ip:30900/`.
+
+Exposing the Alertmanager works in the same fashion, the only difference being,
+that the selector is `alertmanager: <alertmanager-name>`. So the `Alertmanager`
+manifest would look like this:
+
+```
+apiVersion: "monitoring.coreos.com/v1alpha1"
+kind: "Alertmanager"
+metadata:
+  name: "main"
+spec:
+  replicas: 3
+  version: v0.5.1
+  resources:
+    requests:
+      memory: 400Mi
+```
+
+And the `Service` manifest like this:
+
+```
+apiVersion: v1
+kind: Service
+metadata:
+  name: alertmanager-main
+spec:
+  type: NodePort
+  ports:
+  - name: web
+    nodePort: 30903
+    port: 9093
+    protocol: TCP
+    targetPort: web
+  selector:
+    alertmanager: main
+```
+
+And the Alertmanager web UI will be available at `http://node-ip:30903/`.
+
+## Kubernetes API
+
+The Kubernetes API has a feature of forwarding requests from the API to a
+cluster internal `Service`. The general URL scheme to access these is:
+
+```
+http(s)://master-host/api/v1/proxy/namespaces/<namespace>/services/<service-name>:<port-name-or-number>/
+```
+
+> Note for ease of use, you can use `kubectl proxy`, it proxies requests from a
+> local address to the Kubernetes API server and handles authentication for
+> you.
+
+To be able to do so, create a `Service` of type `ClusterIP`:
+
+```
+apiVersion: v1
+kind: Service
+metadata:
+  name: prometheus-main
+spec:
+  type: ClusterIP
+  ports:
+  - name: web
+    port: 9090
+    protocol: TCP
+    targetPort: web
+  selector:
+    prometheus: main
+```
+
+A caveat about this is that Prometheus and Alertmanager need to be configured
+with the full URL they are going to be exposed at. Therefore the `Prometheus`
+manifest will need an entry for `externalUrl`:
+
+```
+apiVersion: "monitoring.coreos.com/v1alpha1"
+kind: "Prometheus"
+metadata:
+  name: "main"
+spec:
+  replicas: 1
+  version: v1.5.0
+  externalUrl: http://127.0.0.1:8001/api/v1/proxy/namespaces/default/services/prometheus-main:web/
+  routePrefix: /
```

Comment:
I was trying to do that initially, however, this is the one case we do actually need to expose this field for. Problem is that the proxy actually rewrites the requests path from `/api/v1/proxy/.../graph` to `/graph`, which wouldn't work with the way the `externalUrl` works internally with Prometheus. When not specifying a `routePrefix` it defaults to what the path segment of the `externalUrl` is, which is wrong here because of this URL rewriting when proxying.